### PR TITLE
EVG 6792: refactor scheduler-related database model

### PR DESCRIPTION
**Pull request by StuartButler**
The  original `distro.planner_settings` fields have been collated into function-specific (sub-)documents: `planner_settings`, `dispatcher_settings` and `host_allocator_settings`. We already have a `finder_settings` sub-document.

Previously, `distro.planner_settings.version` was overloaded. This change allows the 3 stages of scheduling to be independently set to specific versions, e.g. run a _legacy_ `task_finder` with a _tunable_  `task_planner` with a _revised_ `task_dispatcher`.  This can been done using either the API or the _distros_ UI. 

Some minor _db.distros_ schema modifications will be required before merging the final version of this PR.  Afterwards, there will be some redundant document fields to be cleaned up - more details to follow.

There is commented code to be removed and some general clean up outstanding.

### CODE REVIEW

File: model/distro/distro.go [hunk 1]
```diff
@@ -513,21 +560,16 @@ func (d *Distro) GetResolvedPlannerSettings(s *evergreen.Settings) (PlannerSetti
...
 		catcher.Errorf("'%s' is not a valid PlannerSettings.TaskOrdering", resolved.TaskOrdering)
 	}
-
-	// Any validation errors?
 	if catcher.HasErrors() {
 		return PlannerSettings{}, errors.Wrapf(catcher.Resolve(), "cannot resolve PlannerSettings for distro '%s'", d.Id)
 	}
 
-	d.PlannerSettings = resolved
+	d.PlannerSettings = resolved // STU: should we do this?
```
Comment (tychoish) [inline model/distro/distro.go]: yes?

File: model/task_queue_service.go [hunk 2]
```diff
@@ -91,22 +91,30 @@ func (s *taskDispatchService) ensureQueue(distroID string) (CachedDispatcher, er
...
+	//////////////////////////////////////////////////////////////////////////////
+	// config, err := evergreen.GetConfig()
+	// if err != nil {
+	// 	return nil, errors.Wrap(err, "error retrieving Evergreen config")
+	// }
+	// plannerSettings, err := d.GetResolvedPlannerSettings(config)
+	// if err != nil {
+	// 	return nil, errors.Wrapf(err, "error resolving the PlannerSettings for distro '%s'", d.Id)
+	// }
+	//////////////////////////////////////////////////////////////////////////////
```
Comment (tychoish) [inline model/task_queue_service.go]: delete these comments?

File: service/api_task.go [hunk 3]
```diff
@@ -370,8 +370,17 @@ func assignNextAvailableTask(ctx context.Context, taskQueue *model.TaskQueue, di
...
-		case evergreen.PlannerVersionTunable, evergreen.PlannerVersionRevised:
+		// switch d.PlannerSettings.Version {
+		// case evergreen.PlannerVersionTunable, evergreen.PlannerVersionRevised:
+		// 	queueItem, err = dispatcher.RefreshFindNextTask(d.Id, spec)
+		// 	if err != nil {
+		// 		return nil, false, errors.Wrap(err, "problem getting next task")
+		// 	}
+		// default:
+		// 	queueItem = taskQueue.FindNextTask(spec)
+		// }
```
Comment (tychoish) [inline service/api_task.go]: delete

File: service/api_task_test.go [hunk 4]
```diff
@@ -115,13 +115,414 @@ func getStartTaskEndpoint(t *testing.T, as *APIServer, hostId, taskId string) *h
 	return w
 }
 
-func TestAssignNextAvailableTaskWithPlannerSettingVersionLegacy(t *testing.T) {
+// func TestAssignNextAvailableTaskWithPlannerSettingVersionLegacy(t *testing.T) {
```
Comment (tychoish) [inline service/api_task_test.go]: delete or uncomment

**Comment by tychoish** [comment]
is there anything holding up this merge?

**Comment by StuartButler** [comment]
Yes - I need to run a script on the production database before merging.  The script has already been run on staging.

**Comment by StuartButler** [comment]
I'll need to run the script once again on production, prior to deploying.

**Comment by bsamek** [comment]
@StuartButler Can this be merge?

**Comment by StuartButler** [comment]
Not yet. I've decided to add back in 3 fields to `PlannerSettings` to simplify any potential rollback.    Theses fields will then be subsequently removed via another PR.  I'll update this PR shortly and, as discussed, will be deploying this tomorrow morning.

**Pull request merged**